build: fix apk packaging and ABI-versioning

GeorgeSapkin opened the pull request:
Currently there are issues with both ABI-versioned and non ABI-versioned packages. ABI versioned ones [can't be installed side-by-side](https://github.com/openwrt/openwrt/issues/20582) and don't version packaging-related files: `alternatives`, `conffiles`, `conffiles_static`, `list` and `rusers`. And non ABI-versioned ones that provide the same package, [can be installed side-by-side](https://github.com/openwrt/openwrt/issues/20802), but shouldn't be able to.

The updated logic for the APK dependencies and provides is as follows:

- If ABI version is defined:
  - package is named `package_name-ABI_version`
  - package implicitly provides `package_name-ABI_version=package_version`
    this implies that only one version of a package per ABI can be installed at the same time
  - additionally provide `package_name` so multiple packages can be looked up by its base name
  - for each `provides`, provide `provide-ABI_version=package_version`
    this implies that only one version of a provide can be installed at the same time

- else if ABI version is _not_ defined
  - package is named `package_name`
  - package implicitly provides `package_name=package_version`
    this implies that only one version of a package can be installed at the same time
  - if `alternatives` is defined
    - for each `provides`, provide `provide`
      this implies that multiple versions of a provide can be installed at the same time
  - else if `alternatives` is _not_ defined
    - for each `provides`, provide `provide=package_version`
      this implies that only one version of a provide can be installed at the same time

Both cases a package can be looked up by its base name.

ABI version `alternatives`, `conffiles`, `conffiles_static`, `list` and `rusers` files so multiple versions of the same ABI package can be installed side by side, and so they don't overwrite each other's packaging files.

ABI version `EXTRA_DEPENDS` so dependencies can be correctly looked up using the existing OpenWrt semantics without the ABI specified (the big for-loop). This is needed since ABI-versioned libraries no longer provide `package_name=package_version`, so that they can be installed side by side.

> [!CAUTION]
> This throws an error now if `EXTRA_DEPENDS` is being used without constraints, i.e. as `DEPENDS`. Luckily there are only a [few packages doing that](https://github.com/openwrt/packages/pull/27908), and they should be fixed.
>
> Edit: tar is [still an issue](https://github.com/openwrt/packages/pull/27964/).
>
> Edit 2: seconds [attempt to fix tar](https://github.com/openwrt/packages/pull/28114) without breaking anything else.

Remove duplicate dependencies when `EXTRA_DEPENDS` specifies a versioned one that is already in `DEPENDS`.

## ABI is defined

`libsqlite3` has `PROVIDES` set to `libfake` and has two different ABI versions installed. `libfake` is just an example to demonstrate the mechanics, as the library can already be depended upon using e.g. `libsqlite3-0=3.51.0-r1`. Note the ABI-versioned lists.

```
root@OpenWrt:/tmp# apk add --allow-untrusted ./libsqlite3-0-3.51.0-r1.apk
(1/1) Installing libsqlite3-0 (3.51.0-r1)
libsqlite3-0-3.51.0-r1.post-install: Executing script...
OK: 22 MiB in 157 packages

root@OpenWrt:/tmp# apk add --allow-untrusted ./libsqlite3-1-4.00.0-r1.apk
(1/1) Installing libsqlite3-1 (4.00.0-r1)
libsqlite3-1-4.00.0-r1.post-install: Executing script...
OK: 23 MiB in 158 packages

root@OpenWrt:/tmp# apk query --fields name,version,contents,provides libsqlite3-0 libsqlite3-1
Name: libsqlite3-0
Version: 3.51.0-r2
Provides: libfake-0=3.51.0-r2 libsqlite3
Contents:
  lib/apk/packages/libsqlite3-0.list
  usr/lib/libsqlite3.so.0
  usr/lib/libsqlite3.so.3.51.0

Name: libsqlite3-1
Version: 4.00.0-r1
Provides: libfake-1=4.00.0-r1 libsqlite3
Contents:
  lib/apk/packages/libsqlite3-1.list
  usr/lib/libsqlite3.so.1
  usr/lib/libsqlite3.so.4.00.0

root@OpenWrt:/tmp# ls -lh /usr/lib/libsqlite3.so.*
lrwxrwxrwx    1 root     root          20 Nov 20 00:23 /usr/lib/libsqlite3.so.0 -> libsqlite3.so.3.51.0
lrwxrwxrwx    1 root     root          20 Nov 20 00:27 /usr/lib/libsqlite3.so.1 -> libsqlite3.so.4.00.0
-rwxr-xr-x    1 root     root        1.0M Nov  6 18:19 /usr/lib/libsqlite3.so.3.51.0
-rwxr-xr-x    1 root     root        1.0M Nov  6 18:19 /usr/lib/libsqlite3.so.4.00.0
```

## ABI is not defined

Both `avahi-dbus-daemon` and `avahi-nodbus-daemon` provide `avahi-daemon`, but have no ABI specified. This results in `avahi-daemon=0.8-r11` provides for both packages and only one being able to be installed at the same time:

```
root@OpenWrt:/tmp# apk add --allow-untrusted ./avahi-nodbus-daemon-0.8-r11.apk
(1/4) Installing libavahi-nodbus-support (0.8-r10)
libavahi-nodbus-support-0.8-r10.post-install: Executing script...
(2/4) Installing libdaemon (0.14-r5)
libdaemon-0.14-r5.post-install: Executing script...
(3/4) Installing libexpat (2.7.3-r1)
libexpat-2.7.3-r1.post-install: Executing script...
(4/4) Installing avahi-nodbus-daemon (0.8-r11)
avahi-nodbus-daemon-0.8-r11.post-install: Executing script...
23 MiB in 160 packages

root@OpenWrt:/tmp# apk query --fields provides avahi-nodbus-daemon
Provides: avahi-daemon=0.8-r11

root@OpenWrt:/tmp# apk add --allow-untrusted ./avahi-dbus-daemon-0.8-r11.apk
ERROR: unable to select packages:
  avahi-dbus-daemon-0.8-r11:
    conflicts: avahi-nodbus-daemon-0.8-r11[avahi-daemon=0.8-r11]
    satisfies: world[avahi-dbus-daemon><Q1R111s+ke9Vf+eCxDHX2BZVUK54Q=]
  avahi-nodbus-daemon-0.8-r11:
    conflicts: avahi-dbus-daemon-0.8-r11[avahi-daemon=0.8-r11]
    satisfies: world[avahi-nodbus-daemon><Q1BAu7nLI2MgRabpveLTGO2ksQz7E=]
```

### Provides and alternatives

Both `uclient-fetch` and `wget-nossl` provide `wget` and specify alternatives, so provides are not versioned and both packages can be installed at the same time:

```
root@OpenWrt:/tmp# apk query --fields name,version,contents,provides uclient-fetch wget-nossl
Name: uclient-fetch
Version: 2025.10.03~dc909ca7-r1
Provides: wget
Contents:
  bin/uclient-fetch
  lib/apk/packages/uclient-fetch.alternatives
  lib/apk/packages/uclient-fetch.list

Name: wget-nossl
Version: 1.25.0-r1
Provides: gnu-wget wget
Contents:
  lib/apk/packages/wget-nossl.alternatives
  lib/apk/packages/wget-nossl.list
  usr/libexec/wget-nossl

```

Related:
- https://github.com/openwrt/packages/pull/27908
- https://github.com/openwrt/packages/pull/28114

Fixes: https://github.com/openwrt/openwrt/issues/20582
Fixes: https://github.com/openwrt/openwrt/issues/20802

cc: @dangowrt, @efahl, @map-b

efahl commented:
The commit message is spot on.  

> the unversioned virtual packages (e.g., libfake), but not the versioned base package (libsqlite3=3.51.0-r1) (edit: at least that's how I understood this after some digging).

My reading of the apk-tools documentation concurs with yours.

I'm in the middle of upgrading an old Ubuntu server (ugh), so will do a bunch of tests on the code changes after I slog through the upgrade.

efahl commented:
I've been using a list of cases so I can rationalize about this.  See if you think this is complete.  ("Normal" packages don't have any ABI concerns; "ABI" packages are ABI-versioned.)

1) Two or more "normal" packages with same `provides` and should conflict: the `avahi`, `dnsmasq` or `tc` package variants are all examples.  (Solved by `provides:pkg=version ...`)

2) Two or more normal packages with same `provides`, but do not conflict: `uclient-fetch` and `wget-ssl` both provide `wget`, but may be installed in parallel. (Solved by `provides:pseudo-pkg ...`)

3) ABI packages with same provides, but different version, should not conflict.

Other stuff contributing to the mess (adding complications on top of packages like `libustream-*`'s lack of ABI-versioned `.so` files).  This first one needs resolution, too:
```
$ cat /lib/apk/packages/libubus.list  << The list file name should be versioned.
/lib/libubus.so.20251103
```

This next occurrence of  "base name" should not be an issue as it should be isolated to upgrade tools and their specialized use.
```
$ ubus call rpc-sys packagelist  '{"all":true}' | grep libubus
                "libubus": "2025.11.03~0d4bcb56-r1",
```

GeorgeSapkin committed:
build: fix apk packaging and ABI-versioning

The updated logic for the APK dependencies and provides is as follows:

- If ABI version is defined:
  - package is named `package_name-ABI_version`
  - package implicitly provides
    `package_name-ABI_version=package_version`
    this implies that only one version of a package per ABI can be
    installed at the same time
  - additionally provide `package_name` so multiple packages can be
    looked up by its base name
  - for each `provides`, provide `provide-ABI_version=package_version`
    this implies that only one version of a provide can be installed at
    the same time

- else if ABI version is _not_ defined
  - package is named `package_name`
  - package implicitly provides `package_name=package_version`
    this implies that only one version of a package can be installed at
    the same time
  - if `alternatives` is defined
    - for each `provides`, provide `provide`
      this implies that multiple versions of a provide can be installed
      at the same time
  - else if `alternatives` is _not_ defined
    - for each `provides`, provide `provide=package_version`
      this implies that only one version of a provide can be installed
      at the same time

Both cases a package can be looked up by its base name.

ABI version `alternatives`, `conffiles`, `conffiles_static`, `list` and
`rusers` files so multiple versions of the same ABI package can be
installed side by side, and so they don't overwrite each other's
packaging files.

ABI version `EXTRA_DEPENDS` so dependencies can be correctly looked up
using the existing OpenWrt semantics without the ABI specified. This is
needed since ABI-versioned libraries no longer provide
`package_name=package_version`, so that they can be installed side by
side.

Remove duplicate dependencies when `EXTRA_DEPENDS` specifies a versioned
one that is already in `DEPENDS`.

ABI is defined
------------------------------------------------------------------------

`libsqlite3` has `PROVIDES` set to `libfake` and has two different ABI
versions installed. `libfake` is just an example to demonstrate the
mechanics, as the library can already be depended upon using e.g.
`libsqlite3-0=3.51.0-r1`. Note the ABI-versioned lists.

```
root@OpenWrt:/tmp# apk add --allow-untrusted ./libsqlite3-0-3.51.0-r1.apk
(1/1) Installing libsqlite3-0 (3.51.0-r1)
libsqlite3-0-3.51.0-r1.post-install: Executing script...
OK: 22 MiB in 157 packages

root@OpenWrt:/tmp# apk add --allow-untrusted ./libsqlite3-1-4.00.0-r1.apk
(1/1) Installing libsqlite3-1 (4.00.0-r1)
libsqlite3-1-4.00.0-r1.post-install: Executing script...
OK: 23 MiB in 158 packages

root@OpenWrt:/tmp# apk query --fields name,version,contents,provides libsqlite3-0 libsqlite3-1
Name: libsqlite3-0
Version: 3.51.0-r2
Provides: libfake-0=3.51.0-r2 libsqlite3
Contents:
  lib/apk/packages/libsqlite3-0.list
  usr/lib/libsqlite3.so.0
  usr/lib/libsqlite3.so.3.51.0

Name: libsqlite3-1
Version: 4.00.0-r1
Provides: libfake-1=4.00.0-r1 libsqlite3
Contents:
  lib/apk/packages/libsqlite3-1.list
  usr/lib/libsqlite3.so.1
  usr/lib/libsqlite3.so.4.00.0

root@OpenWrt:/tmp# ls -lh /usr/lib/libsqlite3.so.*
lrwxrwxrwx    1 root     root          20 Nov 20 00:23 /usr/lib/libsqlite3.so.0 -> libsqlite3.so.3.51.0
lrwxrwxrwx    1 root     root          20 Nov 20 00:27 /usr/lib/libsqlite3.so.1 -> libsqlite3.so.4.00.0
-rwxr-xr-x    1 root     root        1.0M Nov  6 18:19 /usr/lib/libsqlite3.so.3.51.0
-rwxr-xr-x    1 root     root        1.0M Nov  6 18:19 /usr/lib/libsqlite3.so.4.00.0
```

ABI is not defined
------------------------------------------------------------------------

Both `avahi-dbus-daemon` and `avahi-nodbus-daemon` provide `avahi-daemon`,
but have no ABI specified. This results in `avahi-daemon=0.8-r11` provides
for both packages and only one being able to be installed at the same time:

```
root@OpenWrt:/tmp# apk add --allow-untrusted ./avahi-nodbus-daemon-0.8-r11.apk
(1/4) Installing libavahi-nodbus-support (0.8-r10)
libavahi-nodbus-support-0.8-r10.post-install: Executing script...
(2/4) Installing libdaemon (0.14-r5)
libdaemon-0.14-r5.post-install: Executing script...
(3/4) Installing libexpat (2.7.3-r1)
libexpat-2.7.3-r1.post-install: Executing script...
(4/4) Installing avahi-nodbus-daemon (0.8-r11)
avahi-nodbus-daemon-0.8-r11.post-install: Executing script...
23 MiB in 160 packages

root@OpenWrt:/tmp# apk query --fields provides avahi-nodbus-daemon
Provides: avahi-daemon=0.8-r11

root@OpenWrt:/tmp# apk add --allow-untrusted ./avahi-dbus-daemon-0.8-r11.apk
ERROR: unable to select packages:
  avahi-dbus-daemon-0.8-r11:
    conflicts: avahi-nodbus-daemon-0.8-r11[avahi-daemon=0.8-r11]
    satisfies: world[avahi-dbus-daemon><Q1R111s+ke9Vf+eCxDHX2BZVUK54Q=]
  avahi-nodbus-daemon-0.8-r11:
    conflicts: avahi-dbus-daemon-0.8-r11[avahi-daemon=0.8-r11]
    satisfies: world[avahi-nodbus-daemon><Q1BAu7nLI2MgRabpveLTGO2ksQz7E=]
```

Provides and alternatives
------------------------------------------------------------------------

Both `uclient-fetch` and `wget-nossl` provide `wget` and specify
alternatives, so provides are not versioned and both packages can be
installed at the same time:

```
root@OpenWrt:/tmp# apk query --fields name,version,contents,provides uclient-fetch wget-nossl
Name: uclient-fetch
Version: 2025.10.03~dc909ca7-r1
Provides: wget
Contents:
  bin/uclient-fetch
  lib/apk/packages/uclient-fetch.alternatives
  lib/apk/packages/uclient-fetch.list

Name: wget-nossl
Version: 1.25.0-r1
Provides: gnu-wget wget
Contents:
  lib/apk/packages/wget-nossl.alternatives
  lib/apk/packages/wget-nossl.list
  usr/libexec/wget-nossl

```

Fixes: https://github.com/openwrt/openwrt/issues/20582
Fixes: https://github.com/openwrt/openwrt/issues/20802
Signed-off-by: George Sapkin <george@sapk.in>
Link: https://github.com/openwrt/openwrt/pull/20819
Signed-off-by: Robert Marko <robimarko@gmail.com>

map-b commented:
> I've been using a list of cases so I can rationalize about this. See if you think this is complete. ("Normal" packages don't have any ABI concerns; "ABI" packages are ABI-versioned.)
> 
>     1. Two or more "normal" packages with same `provides` and should conflict: the `avahi`, `dnsmasq` or `tc` package variants are all examples.  (Solved by `provides:pkg=version ...`)
> 
>     2. Two or more normal packages with same `provides`, but do not conflict: `uclient-fetch` and `wget-ssl` both provide `wget`, but may be installed in parallel. (Solved by `provides:pseudo-pkg ...`)
Yes, the ALTERNATIVES mechanism which allow coexistence of "normal" packages with the same provides, which basicly consist in worring about how to avoid overwriting the same files(only binaries?) used by both packages and choosing the best alternative available when installing or unistalling a provider. https://github.com/search?q=repo%3Aopenwrt%2Fopenwrt%20ALTERNATIVES&type=code

Could this ALTERNATIVES be used also for other files than binaries and create symlinks to libraries?

There are packages which would be mix of 1. (same `*origin*` and same provides) and 2. (different `*origin*` and same provides) like ip(ip-full, ip-tiny, tc-tiny, tc-bpf, ...) where they have the same provides but also use the ALTERNATIVES mechanism to allow coexistence of several variants.

> ```
> $ cat /lib/apk/packages/libubus.list  << The list file name should be versioned.
> /lib/libubus.so.20251103
> ```
Nice catch.

map-b commented:
Wow!, such of improve in last pushes @GeorgeSapkin Good work!

BKPepe commented:
@aparcar This is apk related. Can you please take a look?

GeorgeSapkin commented:
@aparcar @robimarko considering 25.12 has recently been forked with apk support, I think this could use a review.

GeorgeSapkin commented:
Actually libgcc seems to have the correct provides:

```
root@OpenWrt:~# apk query --fields name,version,provides libgcc1
Name: libgcc1
Version: 14.3.0-r5
Provides: libgcc
```

So it provides both an ABI-versioned self and a virtual name.

efahl commented:
Ok, I'm going to say ignore it all for now.  

I'll keep watching the smoke tests over on the docker builder and report back here, good or bad, so we can either get closure or do any needed investigations...

robimarko commented:
Yeah, that seems to be the case, but for some weird reason it does not happen on all targets?

GeorgeSapkin commented:
The package in the imagebuilder also seems correct:

```bash
/home/george/Dev/openwrt-sdks/imagebuilder-realtek-rtl839x-20251223/staging_dir/host/bin/apk --root /home/george/Dev/openwrt-sdks/imagebuilder-realtek-rtl839x-20251223/build_dir/target-mips_24kc_musl/root-realtek --keys-dir /home/george/Dev/openwrt-sdks/imagebuilder-realtek-rtl839x-20251223/keys --no-logfile --preserve-env --repositories-file /home/george/Dev/openwrt-sdks/imagebuilder-realtek-rtl839x-20251223/repositories --repository /home/george/Dev/openwrt-sdks/imagebuilder-realtek-rtl839x-20251223/packages/packages.adb --cache-dir /home/george/Dev/openwrt-sdks/imagebuilder-realtek-rtl839x-20251223/dl info -a libgcc
libgcc1-14.3.0-r5 description:
GCC support library

libgcc1-14.3.0-r5 webpage:
http://gcc.gnu.org/

libgcc1-14.3.0-r5 installed size:
128 KiB

libgcc1-14.3.0-r5 depends on:

libgcc1-14.3.0-r5 provides:
libgcc

libgcc1-14.3.0-r5 has auto-install rule:

libgcc1-14.3.0-r5 license:
GPL-3.0-with-GCC-exception
```

GeorgeSapkin commented:
At the very least, replacing libgcc with libgcc1 makes `make manifest` work.

GeorgeSapkin commented:
Or a better question would be - where do base-files inside the imagebuilder come from?

GeorgeSapkin commented:
Somethings off. I'm building the toolchain locally and getting:

```
libc fused dependencies: libgcc
libgcc fused dependencies: 
libatomic fused dependencies: libgcc1
```

Note the libc depending on libgcc instead of libgcc1 even though both libs have the same ` DEPENDS:=+libgcc`. It's as if it's using the ABI of the package and applying it to the dependency.

GeorgeSapkin commented:
I think I have a hunch...
